From d10363e603e43d628016feb5753d760f353b7284 Mon Sep 17 00:00:00 2001
Message-Id: <d10363e603e43d628016feb5753d760f353b7284.1443188573.git.erack@redhat.com>
From: Eike Rathke <erack@redhat.com>
Date: Fri, 25 Sep 2015 13:42:04 +0200
Subject: [PATCH] fix IterateMatrix for ifPRODUCT and ifSUMSQ
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="------------erAck-patch-parts"

This is a multi-part message in MIME format.
--------------erAck-patch-parts
Content-Type: text/plain; charset=UTF-8; format=fixed
Content-Transfer-Encoding: 8bit


... if more than one argument is passed to PRODUCT() or SUMSQ() and
matrix/array arguments are involved the functions returned wrong
results.

For example, wrong behavior
=PRODUCT({2,3},{4,5})   gave 6
=SUMSQ({2,3},{4,5})     gave 34

Correct is
=PRODUCT({2,3},{4,5})   result 120
=SUMSQ({2,3},{4,5})     result 54

Change-Id: Iec7b136a5cc678416f28b1abfe19cd0901ef22b6
(cherry picked from commit c1780ad265ce5f6abb3b91a4f70bf2b3d7a47149)
---
 sc/source/core/tool/interpr6.cxx | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)


--------------erAck-patch-parts
Content-Type: text/x-patch; name="0001-fix-IterateMatrix-for-ifPRODUCT-and-ifSUMSQ.patch"
Content-Transfer-Encoding: 8bit
Content-Disposition: attachment; filename="0001-fix-IterateMatrix-for-ifPRODUCT-and-ifSUMSQ.patch"

diff --git a/sc/source/core/tool/interpr6.cxx b/sc/source/core/tool/interpr6.cxx
index 9ae88f9..14939c2 100644
--- a/sc/source/core/tool/interpr6.cxx
+++ b/sc/source/core/tool/interpr6.cxx
@@ -387,7 +387,7 @@ void IterateMatrix(
         case ifPRODUCT:
         {
             ScMatrix::IterateResult aRes = pMat->Product(bTextAsZero);
-            fRes = aRes.mfFirst;
+            fRes *= aRes.mfFirst;
             fRes *= aRes.mfRest;
             rCount += aRes.mnCount;
         }
@@ -395,6 +395,7 @@ void IterateMatrix(
         case ifSUMSQ:
         {
             ScMatrix::IterateResult aRes = pMat->SumSquare(bTextAsZero);
+            fRes += aRes.mfFirst;
             fRes += aRes.mfRest;
             rCount += aRes.mnCount;
         }

--------------erAck-patch-parts--


